Multi waypoint mission option

breadoven opened the pull request:
PR adds the ability to load multiple WP missions together from the CLI. Once loaded and saved from the CLI an individual mission can be selected using the `nav_multi_waypoint_mission_index` setting with the mission loaded in the usual way prior to flight. `nav_multi_waypoint_mission_index` can be changed using OSD CMS input.

Example CLI WP input:
```
wp 0 1 lat lon alt 0 0 0 0
wp 1 1 lat lon alt 0 0 0 0
wp 2 1 lat lon alt 0 0 0 0
wp 3 4 0 0 0 0 0 0 165
wp 0 1 lat lon alt 0 0 0 0
wp 1 1 lat lon alt 0 0 0 0
wp 2 1 lat lon alt 0 0 0 0
wp 3 1 lat lon alt 0 0 0 0
wp 4 4 0 0 0 1 0 0 165
wp 0 1 lat lon alt 0 0 0 0
wp 1 1 lat lon alt 0 0 0 0
wp 2 1 lat lon alt 0 0 0 165
```
The above WP entries are input to the CLI successively and then saved using the `WP save` command as usual. If `multi_waypoint_mission_index`  is set to 2 then the following mission will be loaded:
```
wp 0 1 lat lon alt 0 0 0 0
wp 1 1 lat lon alt 0 0 0 0
wp 2 1 lat lon alt 0 0 0 0
wp 3 1 lat lon alt 0 0 0 0
wp 4 4 0 0 0 1 0 0 165
```
PR also adds a WP mission entry to the Arming screen, after the Craft name, showing the selected mission number. total missions available and the number of WPs in the selected mission. If the selected mission number is higher than the number of missions then no mission is loaded and the Arming screen mission message indicates the selected mission is out of range. `nav_multi_waypoint_mission_index` resets to 1 when new missions are loaded and saved in the CLI.

breadoven committed:
Initial build

breadoven committed:
Update navigation.c

Make all multi mission WPs visible  in CLI mode

breadoven committed:
Merge branch 'master' into abo_multi_wp_mission_option

breadoven committed:
single WP mission fix

Improve handling single WP missions. Block RTH single WP missions from causing problem finding end of multi mission entry.

breadoven committed:
Update osd.c

Fix arm screen message

breadoven committed:
Merge branch 'master' into abo_multi_wp_mission_option

avsaase commented on src/main/fc/settings.yaml:
```diff
@@ -2253,6 +2253,11 @@ groups:
         max: 65000
+      - name: nav_multi_waypoint_mission_index
+        description: "Index of mission selected from multi mission WP entry. 1 is the first WP mission in the entry. If index exceeds number of missions in the entry then no mission is loaded."
+        default_value: "1"
```
this should be `default_value: 1`, so without the double-quotes.

breadoven replied:
Yep I know. Various settings need updating in line with the new default settings. Already fixed my test version, just need to update affected PRs.

breadoven committed:
Update settings.yaml

breadoven committed:
Add OSD mission field

breadoven committed:
Merge branch 'master' into abo_multi_wp_mission_option

breadoven committed:
Update Settings.md

breadoven committed:
Merge remote-tracking branch 'upstream/master' into abo_multi_wp_mission_option

breadoven committed:
Improvements

- Check number missions loaded during boot with option to load selected mission at the same time
- Mission selection from OSD mission field using stick commands
- Added total number WPs in all missions

breadoven committed:
Update Settings.md

breadoven committed:
Reduce OSD mission field size when armed

breadoven committed:
Fix setting

breadoven committed:
Bug fixes

breadoven committed:
Merge remote-tracking branch 'upstream/master' into abo_multi_wp_mission_option

breadoven committed:
Merge remote-tracking branch 'upstream/master' into abo_multi_wp_mission_option

breadoven committed:
Update settings.yaml

breadoven committed:
Update Settings.md

breadoven committed:
Update navigation.c

breadoven committed:
Merge remote-tracking branch 'upstream/master' into abo_multi_wp_mission_option

breadoven committed:
Update Settings.md

breadoven committed:
Fix reset multi mission index on save

breadoven committed:
Merge remote-tracking branch 'upstream/master' into abo_multi_wp_mission_option

breadoven committed:
Fix logic when mission unloaded

breadoven committed:
Fix USE_NAV define

breadoven committed:
Update osd.c

breadoven committed:
Update navigation.c

breadoven committed:
Merge remote-tracking branch 'upstream/master' into abo_multi_wp_mission_option

breadoven committed:
Update osd.h

breadoven committed:
add MSP control

breadoven committed:
changed method

Changed to load all missions always and only load selected mission on ARM

breadoven committed:
fixes

breadoven committed:
Update navigation.c

protect against incorrect mission index setting

breadoven committed:
Merge remote-tracking branch 'upstream/master' into abo_multi_wp_mission_option

breadoven committed:
revamp

breadoven committed:
update

breadoven committed:
update

Inhibit stick selection for single mission.
Correct setting description.

breadoven committed:
update comments

stronnag reviewed (approved):
I've tested this extensively (via adding support to mwp) and have had no issues.
It seems a bit "inelegant", but as we have no other solution without fine control of config flash I/O, I  see no other or better solution. And this has the benefit of conserving resources on the FC side.
Monumental piece of work pulling all this together.
